src/components/views/admin/BuilderView.tsx

```diff
@@ -39,52 +41,121 @@ import KeyWords from '../../common/settings/KeyWords';
-    value: DEFAULT_TEXT_RESOURCE_SETTING.text,
-    dataKey: DATA_KEYS.TEXT,
-  },
-  [INITIAL_CHATBOT_PROMPT_SETTING_KEY]: {
-    value: DEFAULT_TEXT_RESOURCE_SETTING.text,
-    dataKey: DATA_KEYS.TEXT,
-  },
-  [KEYWORDS_SETTING_KEY]: {
-    value: [] as Keyword[],
-    dataKey: DATA_KEYS.KEYWORDS,
-  },
-};
-type SettingKey = keyof typeof defaultSettings;
-type SettingValue = (typeof defaultSettings)[SettingKey]['value'];
-
-const settingKeys = Object.keys(defaultSettings).map((k) => k as SettingKey);
+const AUTO_SAVE_DEBOUNCE_MS = 700;
+const REFRESH_SAVE_TIME_MS = 20_000;
 
 const BuilderView: FC = () => {
   const { t } = useTranslation();
-  const [settings, setSettings] = useState(defaultSettings);
+  const { lang } = useLocalContext();
+  const {
+    appSettingArray,
+    isError,
+    isLoading,
+    isSuccess,
+    patchAppSetting,
+    postAppSetting,
+  } = useAppSettingContext();
+  const isOnline = useOnlineStatus();
 
+  // This map is used to manage the auto save of each setting.
+  const autoSaveDebounceMap = useRef(new Map<string, DebouncedFunction>());
   // This state is used to avoid to erase changes if another setting is saved.
   const [isClean, setIsClean] = useState(true);
+  const [settings, setSettings] = useState(defaultSettings);
+
+  const lastSavedTime = useRef<Date>();
+  const [lastSavedMsg, setLastSavedMsg] = useState<string>();
+
+  const mostRecentTime = appSettingArray.reduce<Date | null>(
+    getMostRecentTime,
+    null,
+  );
+
+  // This used effect is used to refresh the last saved time periodically.
+  useEffect(() => {
+    const interval = setInterval(() => {
+      const lastTime = lastSavedTime.current;
+      if (lastTime) {
+        setLastSavedMsg(
+          formatDate(lastSavedTime.current?.toString(), {
+            locale: lang ?? DEFAULT_LANG,
+          }),
+        );
+      }
+    }, REFRESH_SAVE_TIME_MS);
+
+    return () => clearInterval(interval);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [lastSavedTime.current]);
+
+  const isChanged = settingKeys
+    .map((settingKey) =>
+      hasKeyChanged({
+        settingKey,
+        setting: settings[settingKey],
+        appSettings: appSettingArray,
+      }),
+    )
+    .some((v) => v);
+
+  const saveSetting = ({
+    settingKey,
+    setting,
+  }: {
+    settingKey: SettingKey;
+    setting: (typeof settings)[typeof settingKey];
+  }): void => {
+    const appSetting = getAppSetting(appSettingArray, settingKey);
+    const { value, dataKey } = setting;
+
+    if (appSetting) {
+      if (
+        !hasKeyChanged({ settingKey, setting, appSettings: appSettingArray })
+      ) {
+        return;
+      }
+
+      patchAppSetting({
+        data: { [dataKey]: value },
+        id: appSetting.id,
+      });
+    } else {
+      postAppSetting({
+        data: { [dataKey]: value },
+        name: settingKey,
+      });
+    }
+
+    if (!isError) {
+      lastSavedTime.current = new Date();
+    }
```

Comment:
Not sure this works as expected. You might reach this line before the post/patch/delete actually finished. 

Also I think in the first place it's not a really good idea to merge all operations' states in the same variables, let's have them separate. (eg. isErrorPostAppSettings)

Reply:
Yes, you're right concerning the timing...